backend-test/TestHelpers.hs

```diff
@@ -40,6 +41,44 @@ runServerPart sp = do
     request <- mockRequest
     simpleHTTP'' sp request
 
+-- | A mock request for running ServerPartWithQuery, specifically for retrieveCourse
+mockRequestWithQuery :: String -> IO Request
+mockRequestWithQuery courseName = do
+    inputsBody <- newMVar []
+    requestBody <- newEmptyMVar
+    return Request
+        { rqSecure          = False
+        , rqMethod          = GET
+        , rqPaths           = ["course"]
+        , rqUri             = "/course"
+        , rqQuery           = "name=" ++ courseName
```

Comment:
It looks here like you're trying to pass the `name` key-value pair in two different ways, likely the result of your experimentation. Building on my comment above, you should tidy up your work by reverting changes to the controller code, and then verifying whether just one of these approaches work, without requiring the other. If you are stuck, you can then push what you've tried and then I'll take a closer look.

Reply:
I don't think you've resolved this comment